osu.Game.Tests/Database/GeneralUsageTests.cs

```diff
@@ -33,6 +36,37 @@ public void TestBlockOperations()
             });
         }
 
+        [Test]
+        public void TestNestedContextCreation()
+        {
+            RunTestWithRealm((realmFactory, _) =>
+            {
+                var mainContext = realmFactory.Context;
+                bool callbackRan = false;
+
+                var subscription = mainContext.All<RealmBeatmap>().SubscribeForNotifications((sender, changes, error) =>
+                {
+                    realmFactory.CreateContext();
+                    callbackRan = true;
+                });
+
+                Task.Factory.StartNew(() =>
+                {
+                    using (var threadContext = realmFactory.CreateContext())
+                    {
+                        threadContext.Write(r => r.Add(new RealmBeatmap(CreateRuleset(), new RealmBeatmapDifficulty(), new RealmBeatmapMetadata())));
+                    }
+                }, TaskCreationOptions.LongRunning | TaskCreationOptions.HideScheduler).Wait();
+
+                // will create a context but also run the callback above (Refresh is implicitly run when getting a new context).
+                realmFactory.CreateContext();
```

Comment:
While I understand that this is testing a usage scenario, could this be tested similarly my last comment? Would greatly improve the readability of this because the addition tasks looks a little wtf.

Also, isn't it possible for this test to run into timing issues if the addition task hasn't run in time before this `realmFactory.CreateContext()` call?

Reply:
It's got a `.Wait()`. How would you write it without the addition task? That's specifically what causes the fail case here.

Reply:
Ah, I missed the `.Wait()`...

What about the following?
```csharp
[Test]
public void TestNestedContextCreation()
{
    RunTestWithRealm((realmFactory, _) =>
    {
        using (var threadContext = realmFactory.CreateContext())
            threadContext.Write(r => r.Add(new RealmBeatmap(CreateRuleset(), new RealmBeatmapDifficulty(), new RealmBeatmapMetadata())));

        bool callbackRan = false;

        using (var ctx = realmFactory.CreateContext())
        {
            var subscription = ctx.All<RealmBeatmap>().SubscribeForNotifications((sender, changes, error) =>
            {
                realmFactory.CreateContext();
                callbackRan = true;
            });

            // Force the callback above to run.
            using (realmFactory.CreateContext())
            {
            }

            subscription.Dispose();
        }

        Assert.IsTrue(callbackRan);
    });
}
```
To me this reads better because:
1. It's not storing the `mainContext` for some reason.
2. It's not doing some inner task to add the beatmap (why is it doing this?)
3. It's sequential - the addition occurs as an init step _before_ everything else.

Reply:
I've updated this to mostly read as you have it.